opendevin/security/invariant/parser.py

```diff
@@ -0,0 +1,154 @@
+
+
+def get_last_id(
+    trace: list[TraceElement],
+) -> Optional[str]:
+    for el in reversed(trace):
+        if type(el) == ToolCall:
+            return el.id
+    return None
+
+
+def parse_action(trace: list[TraceElement], action: Action) -> list[TraceElement]:
+    next_id = get_next_id(trace)
+    inv_trace = []  # type: list[TraceElement]
+    if type(action) == MessageAction:
+        if action.source == EventSource.USER:
+            inv_trace.append(Message(role='user', content=action.content))
+        else:
+            inv_trace.append(Message(role='assistant', content=action.content))
+    elif type(action) == IPythonRunCellAction:
+        function = Function(
+            name='ipython_run_cell',
+            arguments={
+                'code': action.code,
+                'kernel_init_code': action.kernel_init_code,
+            },
+        )
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) == AgentFinishAction:
+        function = Function(name='agent_finish', arguments={'outputs': action.outputs})
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) == CmdRunAction:
+        function = Function(
+            name='cmd_run',
+            arguments={'command': action.command, 'is_confirmed': action.is_confirmed},
+        )
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) == AgentDelegateAction:
+        function = Function(
+            name='agent_delegate',
+            arguments={'agent': action.agent, 'inputs': action.inputs},
+        )
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) == BrowseInteractiveAction:
+        function = Function(
+            name='browse_interactive',
+            arguments={
+                'browser_actions': action.browser_actions,
+                'browsergym_send_msg_to_user': action.browsergym_send_msg_to_user,
+            },
+        )
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) == BrowseURLAction:
+        function = Function(name='browse_url', arguments={'url': action.url})
+        inv_trace.append(Message(role='assistant', content=action.thought))
+        inv_trace.append(ToolCall(id=next_id, type='function', function=function))
+    elif type(action) in [NullAction, ChangeAgentStateAction]:
+        pass
+    else:
+        logger.error(f'Unknown action type: {type(action)}')
+    return inv_trace
```

Comment:
does this function need to get update whenever "Action" definition changes? 🤔 I'm thinking if there's an easy way for us to maintain this? Or trigger a linter error of some sort when someone modified the action but forgot to change here.

Reply:
We could make a method in Actions, `analyzer_trace`, which gets `next_id` and returns the list to append to `inv_trace`. Just thinking it may help, simply because the code will be in front of our eyes in the respective Action subclass.

Related, just wondering: these actions seem to be the current CodeAct/browser actions, e.g. it's not implemented for FileReadAction and others. Will we want all runnable actions or almost all of them? If yes, we could make sure we write definitions, by making the `analyzer_trace` method abstract in Action. This may be overkill, though.

Reply:
Hello, please take a look at the new parser:

https://github.com/OpenDevin/OpenDevin/pull/3058/commits/32c08ffbba9bcfde7d913152c0779b7e9d6407ff#diff-d4f29cbf26d5f92d82b77f0ab8c0d5fa8aeeb1b67f6f4fec0e1f5c7fd04d06f1R51

Changed such that MessageAction is just a user/agent trace. NullAction and ChangeAgentStateAction don't produce traces, and all the other actions are just function calls (the name of each function is as defined in [/opendevin/core/schema/action.py](/OpenDevin/OpenDevin/blob/main/opendevin/core/schema/action.py)

Similar thing for observations.

This should help decouple the analyzer from actions, and custom policies can be written for newer actions as well. (now including FileWrite/Read)

Reply:
Simple and clear, love it.

Thanks so much for doing this! ❤️